File: Tema2/JS/benchmark.js
const { performance } = require('perf_hooks');

// Función que calcula la cantidad de pasos de la conjetura de Collatz para un número
function collatz(n) {
    let steps = 0;
    while (n > 1) {
        if (n % 2 === 0) {
            n = n / 2;
        } else {
            n = 3 * n + 1;
        }
        steps++;
    }
    return steps;
}

// Función principal de benchmarking
function runBenchmark(limit) {
    console.log(`Iniciando benchmark de Conjetura de Collatz para n=${limit}...`);

    // Forzar recolección de basura si se ejecuta con --expose-gc para mediciones más precisas
    if (global.gc) {
        global.gc();
    }

    const memoryBefore = process.memoryUsage();
    const startTime = performance.now();

    // Ejecutar el algoritmo
    for (let i = 1; i <= limit; i++) {
        collatz(i);
    }

    const endTime = performance.now();
    const memoryAfter = process.memoryUsage();

    const timeTaken = endTime - startTime;

    // Cálculo de memoria consumida en MB (Heap Used)
    const memoryUsed = (memoryAfter.heapUsed - memoryBefore.heapUsed) / 1024 / 1024;

    console.log(`\n--- Resultados Benchmarking JavaScript ---`);
    console.log(`Paradigma Dominante: Multiparadigma (Prototípico, Funcional)`);
    console.log(`Mecanismo de Ejecución: JIT (Just-In-Time) / V8 Engine`);
    console.log(`Tiempo de Ejecución: ${timeTaken.toFixed(2)} ms`);
    console.log(`Variación de Consumo de Memoria Heap: ${memoryUsed > 0 ? memoryUsed.toFixed(2) : Math.abs(memoryUsed).toFixed(2)} MB`);

    // Note: Node.js memory management is garbage collected, so heap variation might be negative 
    // or very small for CPU bound tasks without large object allocations.
}

// Se utiliza un N suficientemente grande (n > 50) para generar carga de procesamiento visible
const N = 5000000;
runBenchmark(N);

File: Tema2/Zig/benchmark.zig
const std = @import("std");

const LIMITE: u64 = 1_000_000;

fn collatz_pasos(n: u64) u64 {
    var x: u64 = n;
    var pasos: u64 = 0;
    while (x != 1) {
        if (x % 2 == 0) {
            x = x / 2;
        } else {
            x = 3 * x + 1;
        }
        pasos += 1;
    }
    return pasos;
}

pub fn main() void {
    std.debug.print("=== Benchmarking: Conjetura de Collatz ===\n", .{});
    std.debug.print("Calculando para todos los n < {d}\n\n", .{LIMITE});

    var max_pasos: u64 = 0;
    var n_max_pasos: u64 = 0;
    var i: u64 = 2;

    while (i < LIMITE) : (i += 1) {
        const pasos = collatz_pasos(i);
        if (pasos > max_pasos) {
            max_pasos = pasos;
            n_max_pasos = i;
        }
    }

    std.debug.print("Resultado:\n", .{});
    std.debug.print("  Numero con mas pasos: {d}\n", .{n_max_pasos});
    std.debug.print("  Cantidad de pasos:    {d}\n", .{max_pasos});
}